doc/man/man1/exonerate.1
.\" Exonerate man Page
.TH exonerate 1 "November 2002" exonerate "sequence comparison tool"

.SH NAME
.\"
exonerate \- a generic tool for sequence comparison
.\"

.SH SYNOPSIS
.B exonerate [ options ] <query path> <target path>
.\"

.SH DESCRIPTION
.BR exonerate
is a general tool for sequence comparison.

It uses the
.B C4
dynamic programming library.
It is designed to be both general and fast.
It can produce either gapped or ungapped alignments,
according to a variety of different alignment models.
The C4 library allows sequence alignment using a reduced
space full dynamic programming implementation,
but also allows automated generation of heuristics
from the alignment models, using bounded sparse dynamic programming,
so that these alignments may also be rapidly generated.
Alignments generated using these heuristics will represent
a valid path through the alignment model,
yet (unlike the exhaustive alignments),
the results are not guaranteed to be optimal.
.\"

.RE
.SH CONVENTIONS
.T
A number of conventions (and idiosyncracies) are used within
exonerate.  An understanding of them facilitates interpretation
of the output.
.\"
.TP
.B Coordinates
An in-between coordinate system is used,
where the positions are counted between the symbols,
rather than on the symbols.
This numbering scheme starts from zero.
This numbering is shown below for the sequence "ACGT":
.\"
.SP
.nf

 A C G T
0 1 2 3 4

.SP
.fi
.\"
Hence the subsequence "CG" would have start=1,
end=3, and length=2.
.\"
This coordinate system is used internally in exonerate,
and for all the output formats produced with
the exception of the "human readable" alignment display
and the GFF output where convention and standards dictate
otherwise.
.\"
.TP
.B Reverse Complements
When an alignment is reported on the reverse complement
of a sequence, the coordinates are simply given on
the reverse complement copy of the sequence.
Hence positions on the sequences are never negative.
.\"
Generally, the forward strand is indicated by '+',
the reverse strand by '-', and an unknown or not-applicable
strand (as in the case of a protein sequence) is indicated by '.'
.\"
.TP
.B Alignment Scores
Currently, only the raw alignment scores are displayed.
This score just is the sum of transistion scores
used in the dynamic programming.
For example, in the case of a Smith-Waterman alignment,
this will be the sum of the substitution matrix scores
and the gap penalties.
.\"

.RE
.SH GENERAL OPTIONS
.T
Most arguments have short and long forms.  The long forms
are more likely to be stable over time, and hence should
be used in scripts which call exonerate.
.\"
.TP
.B "\-h | \--shorthelp" <boolean>
Show help.
This will display a concise summary of the available options,
defaults and values currently set.
.\"
.TP
.B "--help" <boolean>
This shows all the help options including the defaults,
the value currently set,
and the environment variable which may be used to set each parameter.
There will be an indication of which options are mandatory.
Mandatory options have no default, and must have a value supplied
for exonerate to run.  If mandatory options are used in order,
their flags may be skipped from the command line (see examples below).
Unlike this man page, the information from this option will always
be up to date with the latest version of the program.
.\"
.TP
.B "\-v | \--version" <boolean>
Display the version number.  Also displays other information
such as the build date and glib version used.
.\"

.RE
.SH SEQUENCE INPUT OPTIONS
Pairwise comparisons will be performed between all query sequences
and all target sequences.
.\"
Generally, for the best performance, shorter sequences
(eg. ESTs, shotgun reads, proteins) should be used as the
query sequences, and longer sequences (eg. genomic sequences)
should be used as the target sequences.
.TP
.B "-q | \--query " <paths>
Specify the query sequences required.  These must be in a FASTA
format file.  Single or muiltiple query sequences may be supplied.
Additionally multiple copies of the fasta file may be supplied
following a --query flag, or by using with multiple --query flags.
.TP
.B "-t | \--target" <paths>
Specify the target sequences required.  Also, must be in a FASTA
format file.  As with the query sequences, single or multiple target
sequences and files may be supplied.
The target filename may by replace by a server name and port number
in the form of
.B hostname:port
when using
.B exonerate-server.
See the man page for
.B exonerate-server
for more information on running exonerate in client:server mode.
.\".B NEW(v2.4.0):
.\"multiple servers may now be used.  These will be queried in parallel
.\"if you have set the
.\".B --cores
.\"option.
.B NEW(v2.4.0):
If an input file is not a FASTA format file,
it is assumed to contain a list of other fasta files,
directories or servers (one per line).
.\"
.TP
.B "-Q | \--querytype" <dna | protein>
Specify the query type to use.  If this is not supplied,
the query type is assumed to be DNA when the first sequence in
the file contains more than 85% [ACGTN] bases.
Otherwise, it is assumed to be peptide.  This option forces the
query type as some nucleotide and peptide sequences
can fall either side of this threshold.
.\"
.TP
.B "-T | \--targettype" <dna | protein>
Specify the target type to use.  The same as
.B --querytype
(above), except that it applies to the target.
.\"
Specifying the sequence type will avoid the overhead
of having to read the first sequence in the database twice
(which may be significant with chromosome-sized sequences)
.\"
.TP
.B "\--querychunkid" <id>
.TP
.B "\--querychunktotal" <total>
.TP
.B "\--targetchunkid" <id>
.TP
.B "\--targetchunktotal" <total>
These options to facilitate running exonerate on compute
farms, and avoid having to split up sequence databases
into small chunks to run on different nodes.
If, for example, you wished to split the target database
into three parts, you would run three exonerate jobs
on different nodes including the options:
.PP
.RS
.PD 0
.TP
--targetchunkid 1 --targetchunktotal 3
.TP
--targetchunkid 2 --targetchunktotal 3
.TP
--targetchunkid 3 --targetchunktotal 3
.PP

NB. The granularity offered by this option only goes
down to a single sequence, so when there are more chunks
than sequences in the database, some processes will do nothing.
.RE
.\"

.TP
.B "\-V | \--verbose" <int>
Be verbose - show information about what is going on during
the analysis.  The default is 1 (little information), the higher
the number given, the more information is printed.
To silence all the default output from exonerate,
use --verbose 0 --showalignment no --showvulgar no
.\"

.SH ANALYSIS OPTIONS
.\"
.TP
.B "\-E | \--exhaustive" <boolean>
Specify whether or not exhaustive alignment should be used.
By default, this is FALSE, and alignment heuristics will be used.
If it is set to TRUE, an exhaustive alignment will be calculated.
This requires quadratic time, and will be much, much slower,
but will provide the optimal result for the given model.
.\"
.TP
.B "\-B | \--bigseq" <int>
Perform alignment of large (multi-megabase) sequences.
This is very memory efficient and fast when both sequences
are chromosome-sized, but currently does not currently permit the use
of a word neighbourhood (ie. exactly matching seeds only).
.\"
.TP
.B "\--revcomp" <boolean>
Include comparison of the reverse complement of the query and target
where possible.  By default, this option is enabled,
but when you know the gene is definitely on the forward strand
of the query and target,
this option can halve the time taken to compute alignments.
.\"
.TP
.B "\--forcescan" <none | query | target>
Force the FSM to scan the query sequence rather than the target.
This option is useful, for example, if you have a single piece
of genomic sequence and you with to compare it to the whole of
dbEST.  By scanning the database, rather than the query,
the analysis will be completed much more quickly, as the overheads
of multiple query FSM construction, multiple target reading
and splice site predictions will be removed.
By default, exonerate will guess the optimal strategy based
on database sequence sizes.
.\"
.TP
.B "\--saturatethreshold" <number>
When set to zero, this option does nothing.
Otherwise, once more than this number of words
(in addition to the expected number of words by chance)
have matched a position on the query, the position
on the query will be 'numbed' (ignore further matches)
for the current pairwise comparison.
.\"
.TP
.B "\--customserver" <command>
When using exonerate in client:server mode with a non-standard
server, this command allows you to send a custom command to the
server.  This command is sent by the client (exonerate)
before any other commands, and is provided as a way of passing
parameters or other commands specific to the custom server.  See the
.B exonerate-server
man page for more information on running exonerate in client:server mode.
.\"
.\".TP
.\".B "\--cores" <number>
.\"The number of cores/CPUs/threads that should be used.
.\"On a multi-core or multi-CPU machine, increasing this ammount
.\"allows alignment computations to run in parallel on separate CPUs/cores.
.\"NB.  Generally, it is better to parallelise the analysis
.\"by splitting it up into separate jobs, but this option may prove
.\"useful for problems such as interactive single-gene queries.
.\"

.SH FASTA DATABASE OPTIONS
.TP
.B "\--fastasuffix" <extension>
If any of the inputs given with
.B --query
or
.B --target
are directories, then exonerate will recursively
descent these directories, reading all files
ending with this suffix as fasta format input.
.\"

.SH GAPPED ALIGNMENT OPTIONS
.TP
.B "\-m | \--model" <alignment model>
Specify the alignment model to use.
The models currently supported are:
.\"
.PP
.RS
.PD 0
.\"
.TP
.B ungapped
The simplest type of model, used by default.
An appropriate model with be selected automatically
for the type of input sequences provided.
.\"
.TP
.B ungapped:trans
This ungapped model includes translation of all frames of both
the query and target sequences.  This is similar to an ungapped
tblastx type search.
.\"
.TP
.B affine:global
This performs gapped global alignment, similar
to the Needleman-Wunsch algorithm, except with affine gaps.
Global alignment requires that both the sequences in their entirety
are included in the alignment.
.\"
.TP
.B affine:bestfit
This performs a best fit or best location alignment
of the query onto the target sequence.  The entire query sequence
will be included in the alignment, but only the best location
for its alignment on the target sequence.
.\"
.TP
.B affine:local
This is local alignment with affine gaps,
similar to the Smith-Waterman-Gotoh algorithm.
A general-purpose alignment algorithm.
As this is local alignment, any subsequence of the query
and target sequence may appear in the alignment.
.\"
.TP
.B affine:overlap
This type of alignment finds the best overlap between the
query and target.  The overlap alignment must include
the start of the query or target
and the end of the query or the target sequence,
to align sequences which overlap at the ends,
or in the mid-section of a longer sequence..
This is the type of alignment frequently used in assembly
algorithms.
.\"
.TP
.B est2genome
This model is similar to the affine:local model,
but it also includes intron modelling on the target sequence
to allow alignment of spliced to unspliced coding sequences
for both forward and reversed genes.  This is similar to the
alignment models used in programs such as EST_GENOME and sim4.
.\"
.TP
.B ner
NERs are non-equivalenced regions - large regions in both
the query and target which are not aligned.  This model can be
used for protein alignments where strongly conserved helix regions
will be aligned, but weakly conserved loop regions are not.
Similarly, this model could be used to look for co-linearly
conserved regions in comparison of genomic sequences.
.\"
.TP
.B protein2dna
This model compares a protein sequence to a DNA sequence,
incorporating all the appropriate gaps and frameshifts.
.\"
.TP
.B protein2dna:bestfit
This is a bestfit version of the protein2dna model,
with which the entire protein is included in the alignment.
It is currently only available when using exhaustive alignment.
.\"
.TP
.B protein2genome
This model allows alignment of a protein sequence to genomic
DNA.   This is similar to the protein2dna model,
with the addition of modelling of introns and intron phases.
This model is similar to those used by genewise.
.\"
.TP
.B protein2genome:bestfit
This is a bestfit version of the protein2genome model,
with which the entire protein is included in the alignment.
It is currently only available when using exhaustive alignment.
.\"
.TP
.B coding2coding
This model is similar to the ungapped:trans model, except
that gaps and frameshifts are allowed.
It is similar to a gapped tblastx search.
.\"
.TP
.B coding2genome
This is similar to the est2genome model, except that the
query sequence is translated during comparison, allowing
a more sensitive comparison.
.\"
.TP
.B cdna2genome
This combines properties of the est2genome and coding2genome
models, to allow modeling of an whole cDNA where a central
coding region can be flanked by non-coding UTRs.
When the CDS start and end is known it may be specified
using the --annotation option (see below)
to permit only the correct coding region to appear in the alignemnt.
.\"
.TP
.B genome2genome
This model is similar to the coding2coding model, except
introns are modelled on both sequences.
(not working well yet)
.\"

.PD
.RE
.\"
.TP
.\"
The short names u, u:t, a:g, a:b, a:l, a:o, e2g, ner,
p2d, p2d:b p2g, p2g:b, c2c, c2g cd2g and g2g
can also be used for specifying models.
.TP
.B "\-s | \--score" <threshold>
This is the overall score threshold.
Alignments will not be reported below this threshold.
For heuristic alignments, the higher this threshold,
the less time the analysis will take.
.\"
.TP
.B "\--percent" <percentage>
Report only alignments scoring at least this percentage
of the maximal score for each query.
eg. use
.B --percent 90
to report alignments with 90% of the maximal
score optainable for that query.
This option is useful not only because it reduces
the spurious matches in the output,
but because it generates query-specific thresholds (unlike
.B --score
) for a set of queries of differing lengths,
and will also speed up the search considerably.
.B NB.
with this option, it is possible to have a cDNA
match its corresponding gene exactly,
yet still score less than 100%,
due to the addition of the intron penalty scores,
hence this option must be used with caution.
.\"
.TP
.B "\--showalignment" <boolean>
Show the alignments in an human readable form.
.\"
.TP
.B "\--showsugar" <boolean>
Display "sugar" output for ungapped alignments.
Sugar is Simple UnGapped Alignment Report, which displays
ungapped alignments one-per-line.  The sugar line starts with
the string "sugar:" for easy extraction from the output,
and is followed by the the following 9 fields in the order below:
.\"
.PP
.RS
.PD 0
.\"
.TP 16
.B query_id
Query identifier
.\"
.TP 16
.B query_start
Query position at alignment start
.\"
.TP 16
.B query_end
Query position alignment end
.\"
.TP 16
.B query_strand
Strand of query matched
.\"
.TP 16
.B target_id
|
.\"
.TP 16
.B target_start
| the same 4 fields
.\"
.TP 16
.B target_end
| for the target sequence
.\"
.TP 16
.B target_strand
|
.\"
.TP 16
.B score
The raw alignment score
.\"
.PD
.RE
.\"
.TP
.B "\--showcigar" <boolean>
Show the alignments in "cigar" format.
Cigar is a Compact Idiosyncratic Gapped Alignment Report,
which displays gapped alignments one-per-line.
The format starts with the same 9 fields as sugar output
(see above), and is followed by a series of <operation, length>
pairs where operation is one of match, insert or delete,
and the length describes the number of times this operation
is repeated.
.\"
.TP
.B "\--showvulgar" <boolean>
Shows the alignments in "vulgar" format.
Vulgar is Verbose Useful Labelled Gapped Alignment Report,
This format also starts with the same 9 fields as sugar output
(see above), and is followed by a series of
<label, query_length, target_length> triplets.
The label may be one of the following:
.\"
.PP
.RS
.PD 0
.\"
.TP
.B M
Match
.\"
.TP
.B C
Codon
.\"
.TP
.B G
Gap
.\"
.TP
.B N
Non-equivalenced region
.\"
.TP
.B 5
5' splice site
.\"
.TP
.B 3
3' splice site
.\"
.TP
.B I
Intron
.\"
.TP
.B S
Split codon
.\"
.TP
.B F
Frameshift
.\"
.PD
.RE
.LP
.\"
.TP
.B "\--showquerygff" <boolean>
Report GFF output for features on the query sequence.
See http://www.sanger.ac.uk/Software/formats/GFF for more information.
.\"
.TP
.B "\--showtargetgff" <boolean>
Report GFF output for features on the target sequence.
.\"
.TP
.B "\--ryo" <format>
Roll-your-own output format.
This allows specification of a printf-esque format
line which is used to specify which information to include
in the output, and how it is to be shown.
The format field may contain the following fields:
.PP
.RS
.PD 0
.TP
.B %[qt][idlsSt]
For either {query,target}, report the
{id,definition,length,sequence,Strand,type}
Sequences are reported in a fasta-format like block (no headers).
.TP
.B %[qt]a[bels]
For either {query,target} region which occurs
.B in the alignment,
report the {begin,end,length,sequence}
.TP
.B %[qt]c[bels]
For either {query,target} region which occurs
in the
.B coding sequence
in the alignment,
report the {begin,end,length,sequence}
.TP
.B %s
The raw score
.TP
.B %r
The rank (in results from a bestn search)
.TP
.B %m
Model name
.TP
.B %e[tism]
Equivalenced {total,id,similarity,mismatches}
(ie. %em == (%et - %ei))
.TP
.B %p[ciIsS]
Percent {query_coverage,id,BLASTId,similarity,Self}
over the equivalenced portions of the alignment.
(i.e. %pi == 100*(%ei / %et)), with the exception of percent BLASTId, which is
calculated as %pI == 100*(%et + query gaps + target gaps).
Percent query coverage is defined as %pc == 100*(%et/%ql).
Percent Self is the score over the equivalenced portions
of the alignment as a percentage of the self comparison score
of the query sequence.
.TP
.B %g
Gene orientation ('+' = forward, '-' = reverse, '.' = unknown)
.TP
.B %S
Sugar block (the 9 fields used in sugar output (see above)
.TP
.B %C
Cigar block (the fields of a cigar line after the sugar portion)
.TP
.B %V
Vulgar block (the fields of a vulgar line after the sugar portion)
.TP
.B %%
Expands to a percentage sign (%)
.TP
.B \en
Newline
.TP
.B \et
Tab
.TP
.B \e\e
Expands to a backslash (\e)
.TP
.B \e{
Open curly brace
.TP
.B \e}
Close curly brace
.TP
.B {
Begin per-transition output section
.TP
.B }
End per-transition output section
.TP
.B %P[qt][sabe]
Per-transition output for {query,target} {sequence,advance,begin,end}
.TP
.B %P[nsl]
Per-transition output for {name,score,label}
.PD
.RE
.PP
This option is very useful and flexible.  For example,
to report all the sections of query sequences which feature
in alignments in fasta format, use:
.PP
--ryo \fB">%qi %qd\en%qas\en"\fP
.P
To output all the symbols and scores in an alignment,
try something like:
.PP
--ryo \fB"%V{%Pqs %Pts %Ps\en}"\fP
.LP
.\"
.TP
.B "\-n | \--bestn" <number>
Report the best N results for each query.
(Only results scoring better than the score threshold
 will be reported).
The option reduces the amount of output generated,
and also allows exonerate to speed up the search.
.\"
.TP
.B "\-S | \--subopt" <boolean>
This option allows for the reporting of (Waterman-Eggert style)
suboptimal alignments.
(It is on by default.)
All suboptimal (ie. non-intersecting) alignments will
be reported for each pair of sequences scoring
at least the threshold provided by
.B --score.

When this option is used with exhaustive alignments,
several full quadratic time passes will be required,
so the running time will be considerably increased.
.\"
.TP
.B "\-g | \--gappedextension" <boolean>
Causes a gapped extension stage to be performed
ie. dynamic programming is applied in arbitrarily shaped
and dynamically sized regions surrounding HSP seeds.
The extension threshold is controlled by the --extensionthreshold
option.

Although sometimes slower than BSDP,
gapped extension improves sensitivity with weak,
gap-rich alignments such as during cross-species comparison.

.B NB. This option is now the default.  Set it to false
to reverse to the old BSDP type alignments.
This option may be slower than BSDP for some large scale analyses
with simple alignment models.
.\"
.TP
.B "\--refine" <strategy>
Force exonerate to refine alignments generated
by heuristics using dynamic programming over larger regions.
This takes more time, but improves the quality of the final
alignments.

The strategies available for refinement are:
.\"
.PP
.RS
.PD 0
.\"
.TP
.B none
The default - no refinement is used.
.\"
.TP
.B full
An exhaustive alignment is calculated from the pair of sequences
in their entirety.
.\"
.TP
.B region
DP is applied just to the region of the sequences covered
by the heuristic alignment.
.\"
.PD
.RE
.LP
.\"
.TP
.B "\--refineboundary" <size>
Specify an extra boundary to be included in the region
subject to alignment during refinement by region.
.\"
.RE

.SH VITERBI ALGORITM OPTIONS
.TP
.B "\-D | \--dpmemory" <Mb>
The exhaustive alignment traceback routines use a Hughey-style
reduced memory technique.  This option specifies how much memory
will be used for this.  Generally, the more memory is permitted
here, the faster the alignments will be produced.
.\"

.SH CODE GENERATION OPTIONS
.TP
.B "\-C | \--compiled" <boolean>
This option allows disabling of generated code for dynamic programming.
It is mainly used during development of exonerate.
When set to FALSE, an "interpreted" version of the dynamic programming
implementation is used, which is much slower.
.\"

.SH HEURISTIC OPTIONS
.\"
.PD 0
.P
.B "\--terminalrangeint"
.P
.B "\--terminalrangeext"
.P
.B "\--joinrangeint"
.P
.B "\--joinrangeext"
.P
.B "\--spanrangeint"
.P
.TP
.B "\--spanrangeext"
These options are used to specify the size of the sub-alignment
regions to which DP is applied around the ends of the HSPs.
This can be at the HSP ends (terminal range), between HSPs
(join range), or between HSPs which may be connected by a large
region such as an intron or non-equivalenced region (span range).
These ranges can be specified for a number of matches back
onto the HSP (internal range) or out from the HSP (external range).
.PD

.SH SEEDED DYNAMIC PROGRAMMING OPTIONS
.TP
.B "\-x | \--extensionthreshold" <score>
This is the amount by which the score will be allowed
to degrade during SDP.
This is the equivalent of the hspdropoff penalties,
except it is applied during dynamic programming, not HSP extension.
Decreasing this parameter will increase the speed of the SDP,
and increasing it will increase the sensitivity.
.\"
.TP
.B "\--singlepass " <boolean>
By default the suboptimal SDP alignments are reported by
a singlepass algorithm, but may miss some suboptimal
alignments that are close together.
This option can be used to force the use of a multipass
suboptimal alignment algorithm for SDP,
resulting in higher quality suboptimal alignments.
.\"
.SH BSDP OPTIONS
.\"
.TP
.B "\--joinfilter" <limit>
(experimental)

Only allow consider this number of SARs for
joining HSPs together.  The SARs with the highest potential
for appearing in a high-scoring alignment are considered.
This option useful for limiting time and memory usage
when searching unmasked data with repetitive sequences,
but should not be set too low, as valid matches may be ignored.
Something like
.B --joinfilter 32
seems to work well.
.\"
.SH SEQUENCE OPTIONS
.\"
.TP
.B "\--annotation" <path>
Specify basic sequence annotation information.
This is most useful with the cdna2genome model,
but will work with other models.
The annotation file contains four fields per line:
.PP
.RS
.PD 0
.TP
<id> <strand> <cds_start> <cds_length>

.TP
Here is a simple example of such a file for 4 cDNAs:

.PP
.TP
dhh.human.cdna + 308 1191
.P
dhh.mouse.cdna + 250 1191
.P
csn7a.human.cdna + 178 828
.P
csn7a.mouse.cdna + 126 828
.PP
.P

These annotation lines will also work when only the first two fields are used.
This can be used when specifying which strand of a specific sequence
should be included in a comparison.
.\"
.RE

.SH SYMBOL COMPARISON OPTIONS

.TP
.B "\--softmaskquery" <boolean>
Indicate that the query is softmasked.  See description below for
.B --softmasktarget
.\"
.TP
.B "\--softmasktarget" <boolean>

ChangeLog
2.4.0 -> 2.4.1
    o Added %pc to --ryo (percent query coverage)
    o Added %pI to --ryo (BLAST-compatible percent identity)
    o Added GFF2 Target attribute to exon features produced by --showtargetgff
    o Added passing of querytype param to server
    o Fixed rare bug with --refine region
    o Disabled multi-threading due to exonerate-server reliability issues
    o Fixed bugs in non-threaded exonerate-server (configured with
      --disable-pthreads, which is now the default)
    o Replaced several deprecated glib functions with POSIX equivalents
    o Fixed Selenocysteine (U) parsing.
      As with BLAST+ 2.4.0, exonerate scores Selenocysteine as Cysteine (C)

2.3.0 -> 2.4.0
    o Fixed many thread safety bugs
    o Fixed bugs with large files over 4GB
    o Modified client to work with multiple servers
    o Fixed server to handle SIGPIPE
    o Added support for FOSN input (can be files, directories or servers)
    o Fixed bug with --wordambiguity and --wordjump (with normal FSM)
    o Added %pS to --ryo (percent self score over equivalenced regions)

2.2.0 -> 2.3.0
    o Fixed several bugs with thread safety for exonerate-server
    o Fixed bug with custom genetic codes not working
    o Fixed bug with overflow of seedrepeat horizon in hspset
    o Added passing of wordlimit params to server
    o Implemented --cores option for multi-threaded alignment
    o Added identity and similarity score for GFF genes and exons
    o Increased default FSM memory limit to 256Mb
    o Added --revcomp option to disable searching of reverse complement
    o Fixed bug with compact-FSM (VFSM) based database searching
    o Added --wordambiguity option for searching ambiguous reference seqs
    o Fixed bug with esd2esi and .esd files >2Gb

2.1.0 -> 2.2.0
    o Fixed bug with esd2esi when softmasking is off
    o Removed all GMemChunk usage due to caching problems with glib-2
    o Fixed various minor memory leaks

2.0.0 -> 2.1.0
    o Fixed portability bugs for x86_64
    o Added linecount: tag for long client:server messages
    o Added setsockopt(TCP_NODELAY) to improved socket performance
    o Changed server to use pthreads by default to avoid LSF memlimit problems
    o Added faster implementation of Sequence_strncpy()
    o Fixed bug with sequence caching
    o Added -V 3 client-server debug code
    o Prevented sub-optimial alignments below geneseed threshold from appearing
    o Fixed sorting of translated indices
    o Completed implementation of server-side geneseed
    o Fixed bug with .esd file parsing
    o Made compilation use glib2 by default
    o Fixed SDP memory management bug
    o Fixed segfault bug with --refine (Thanks to Don Gilbert)
    o Improved splice site scoring
    o Reverted to protein substition scoring for cd2g model
    o Fixed bug with fasta parser and \r\n type newlines

1.4.0 -> 2.0.0
    o Modified exonerate to work in Client:Server mode
    o Fixed --refine to work with SDP alignments
    o Disabled codegen warnings from bootstrapper during compilation
    o Added --geneticcode option for using alternative genetic codes
    o Added --splice5 and --splice3 to allow alternative splice site PSSMs
    o Fixed several bugs when scanning query sequences (eg. --forcescan query)
    o Fixed to report query on forward strand whenever possible.
    o Fixed --ryo sequence dumping with --bestn
    o Fixed a bug with missing seeds when using --annotation
    o Changed license to GPLv3
    o Added protein2dna:bestfit and protein2genome:bestfit models
      (works with exhaustive alignment only)

1.3.0 -> 1.4.0
    o Improved splice site cache memory management
    o Added --geneseed option to speed up large-scale analysis
    o Fixed a bug with 3'utr coordinates in GFF output
    o Fixed a crash during GFF dumping when using glib2
    o Added error message for failed seeks when input is stdin

1.2.0 -> 1.3.0
    o Fixed a bad memory leak in boundary.c
    o Fixed bug with selenocysteines in query sequences
    o Fixed bug with subsequence creation (and --ryo problems)
    o Fixed a bug with transition mapping in reduced space alignments
    o Removed variable_submat to fix bugs with codegen and allow speed ups
    o Fixed bug with --usetlaaa and translated models
    o Fixed emit calculation in HPair for BSDP
    o Added C field to vulgar for disambiguation of codon matches
    o Fixed bug with region finding in exhaustive DP codegen
    o Fixed GFF bugs (start always < end) and off-by-one bug on rev strand
    o Fixed bugs with codegen for cd2g
    o Fixed bug with chained silent transitions in reduced space traceback
    o Fixed bug broken bigseq comparisons
    o Fixed --percent and --bestn ungapped comparisons

1.1.0 -> 1.2.0
    o Fixed bugs with SDP and short exons in spanned models
    o Fixed bug with overwriting of some valid SDP span scores
    o Fixed forward coordinates in GFF
    o Added utr3,utr5 and cds fields to GFF output
    o Fixed rounding bug in codonsubmat
      (this prevented codon models working on some machines)
    o Updated SDP codegen
    o Fixed compilation to work with both glib-1 or glib-2
    o Fixed phase macros
    o Fixed build problems on OSF
    o Added initial client/server code (nothing useful yet) 

1.0.0 -> 1.1.0
    o Replaced C4_Model_find_*() functions using names
    o Fixed DP bug with non-local models
    o Added Seeder and Match modules to support multi-hspset models
    o Added %tcs etc to --ryo for dumping coding sequences
    o Added CodonSubmat module
    o Changed WordHood to work in codon space
    o Changed Intron and Phase models to support joint introns for g2g
    o Fixed bug with display of selenocysteine containing alignments
    o Added --annotation option for cd2g model
    o Fixed bug with introns in cigar strings
    o Modified configure.in to compile cleanly on solaris
    o Changed to use abstract sequences (requires optimisations)
    o Fixed bug with --bestn tiebreaking

src/program/exonerate.c
/****************************************************************\
*                                                                *
*  exonerate : a generic sequence comparison tool                *
*                                                                *
*  Guy St.C. Slater..   mailto:guy@ebi.ac.uk                     *
*  Copyright (C) 2000-2009.  All Rights Reserved.                *
*                                                                *
*  This source code is distributed under the terms of the        *
*  GNU General Public License, version 3. See the file COPYING   *
*  or http://www.gnu.org/licenses/gpl.txt for details            *
*                                                                *
*  If you use this code, please keep this notice intact.         *
*                                                                *
\****************************************************************/

#include "argument.h"
#include "analysis.h"
#include "gam.h"
#include "optimal.h"
#include "codegen.h"
#include "fastadb.h"

#include "est2genome.h"
#include "affine.h"
#include "ner.h"
#include "intron.h"
#include "heuristic.h"
#include "protein2dna.h"
#include "frameshift.h"
#include "alphabet.h"
#include "hspset.h"
#include "match.h"
#include "alignment.h"
#include "sar.h"
#include "sdp.h"
#include "splice.h"

int Argument_main(Argument *arg){
    register Analysis *analysis;
    register ArgumentSet *as_input =
        ArgumentSet_create("Sequence Input Options");
    GPtrArray *query_path_list, *target_path_list;
    Alphabet_Type query_type, target_type;
    gint query_chunk_id, target_chunk_id,
         query_chunk_total, target_chunk_total,
         verbosity;
    /**/
    ArgumentSet_add_option(as_input, 'q', "query", "path",
    "Specify query sequences as a fasta format file", NULL,
    NULL, &query_path_list);
    ArgumentSet_add_option(as_input, 't', "target", "path",
    "Specify target sequences as a fasta format file", NULL,
    NULL, &target_path_list);
    /**/
    ArgumentSet_add_option(as_input, 'Q', "querytype",
    "alphabet type", "Specify query alphabet type", "unknown",
    Alphabet_Argument_parse_alphabet_type, &query_type);
    ArgumentSet_add_option(as_input, 'T', "targettype",
    "alphabet type", "Specify target alphabet type", "unknown",
    Alphabet_Argument_parse_alphabet_type, &target_type);
    /**/
    ArgumentSet_add_option(as_input, '\0', "querychunkid", NULL,
    "Specify query job number", "0",
    Argument_parse_int, &query_chunk_id);
    ArgumentSet_add_option(as_input, '\0', "targetchunkid", NULL,
    "Specify target job number", "0",
    Argument_parse_int, &target_chunk_id);
    ArgumentSet_add_option(as_input, '\0', "querychunktotal", NULL,
    "Specify total number of query jobs", "0",
    Argument_parse_int, &query_chunk_total);
    ArgumentSet_add_option(as_input, '\0', "targetchunktotal", NULL,
    "Specify total number of target jobs", "0",
    Argument_parse_int, &target_chunk_total);
    /**/
    ArgumentSet_add_option(as_input, 'V', "verbose", "level",
    "Show search progress", "1",
    Argument_parse_int, &verbosity);
    /**/
    Argument_absorb_ArgumentSet(arg, as_input);
    Translate_ArgumentSet_create(arg);
    Analysis_ArgumentSet_create(arg);
    FastaDB_ArgumentSet_create(arg);
    GAM_ArgumentSet_create(arg);
    Viterbi_ArgumentSet_create(arg);
    Codegen_ArgumentSet_create(arg);
    Heuristic_ArgumentSet_create(arg);
    SDP_ArgumentSet_create(arg);
    BSDP_ArgumentSet_create(arg);
    /**/
    Sequence_ArgumentSet_create(arg);
    Match_ArgumentSet_create(arg);
    Seeder_ArgumentSet_create(arg);
    Affine_ArgumentSet_create(arg);
    NER_ArgumentSet_create(arg);
    Intron_ArgumentSet_create(arg);
    Frameshift_ArgumentSet_create(arg);
    Alphabet_ArgumentSet_create(arg);
    HSPset_ArgumentSet_create(arg);
    Alignment_ArgumentSet_create(arg);
    SAR_ArgumentSet_create(arg);
    Splice_ArgumentSet_create(arg);
    /**/
    Argument_process(arg, "exonerate",
      "A generic sequence comparison tool\n"
      "Guy St.C. Slater. guy@ebi.ac.uk. 2000-2009.\n",
      "\n"
      "Examples of use:\n"
      "\n"
      "1. Ungapped alignment of any DNA or protein sequences:\n"
      "    exonerate queries.fa targets.fa\n"
      "2. Gapped alignment of Mouse proteins to Fugu proteins:\n"
      "    exonerate --model affine:local mouse.fa fugu.fa\n"
      "3. Find top 10 matches of each EST to a genome:\n"
      "    exonerate --model est2genome --bestn 10 est.fa genome.fa\n"
      "4. Find proteins with at least a 50% match to a genome:\n"
      "    exonerate --model protein2genome --percent 50 p.fa g.fa\n"
      "5. Perform a full Smith-Waterman-Gotoh alignment:\n"
      "    exonerate --model affine:local --exhaustive yes a.fa b.fa\n"
      "6. Many more combinations are possible.  To find out more:\n"
      "    exonerate --help\n"
      "    man exonerate\n"
      "\n");
    if(verbosity > 0)
        Argument_info(arg);
    /**/
    analysis = Analysis_create(query_path_list, query_type,
                               query_chunk_id, query_chunk_total,
                               target_path_list, target_type,
                               target_chunk_id, target_chunk_total,
                               verbosity);
    Analysis_process(analysis);
    Analysis_destroy(analysis);
    /**/
    if(verbosity > 0)
        g_print("-- completed exonerate analysis\n");
    return 0;
    }

/**/

